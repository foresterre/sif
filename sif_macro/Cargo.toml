[package]
name = "sif_macro"
version = "0.1.0"
authors = ["Martijn Gribnau <garm@ilumeo.com>"]
edition = "2018"
license = "MIT OR Apache-2.0"
description = "Part of Sif: attribute macro for parameterized tests"
documentation = "https://docs.rs/sif_macro"
repository = "https://github.com/foresterre/sif"
readme = "../README.md"
keywords = ["parameterized", "attribute", "test", "unit-test", "junit"]
categories = ["development-tools", "development-tools::testing"]
autotests = false

[lib]
proc-macro = true

[[test]]
name = "trybuild_cases"
path = "tests/trybuild.rs"

[[test]]
name = "trycompile_cases"
path = "tests/trycompile.rs"

[dependencies]
proc-macro2 = "1"
quote = "1"
syn = { version = "1", features = ["full", "extra-traits"] }

[dev-dependencies]
trybuild = "1.0"
trycompile = { path = "../trycompile" }