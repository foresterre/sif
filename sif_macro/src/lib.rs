#[macro_use]
extern crate syn;
extern crate proc_macro;

use crate::case_parser::CasesFn;
use quote::quote;
use syn::{Attribute, Block, Expr, Ident, Type, Visibility};

mod case_parser;

#[proc_macro_attribute]
pub fn parameterized(
    _args: proc_macro::TokenStream,
    fn_body: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let inputs = parse_macro_input!(fn_body as CasesFn);

    let visibility = inputs.fn_visibility();
    let mod_ident = inputs.fn_ident();
    let other_attributes = inputs.regular_attrs();
    let test_parameters = inputs.fn_parameters();
    let arguments = inputs.case_attrs();
    let body = inputs.fn_body();
    let test_cases = arguments
        .iter()
        .map(|exprs| exprs.values.iter().collect::<Vec<&Expr>>())
        .enumerate()
        .map(|(nth, exprs)| {
            create_test_case(
                Ident::new(&format!("case_{}", nth), inputs.fn_span()),
                &test_parameters,
                &exprs,
                body,
                visibility,
                &other_attributes,
            )
        });

    (quote! {
        #[cfg(test)]
        #visibility mod #mod_ident {
            use super::*;

            #(#test_cases)*
        }

    })
    .into()
}

fn create_test_case(
    ident: Ident,
    params: &[(&Ident, &Type)],
    exprs: &[&Expr],
    body: &Block,
    vis: &Visibility,
    attributes: &[&Attribute],
) -> proc_macro2::TokenStream {
    assert_eq!(
        params.len(),
        exprs.len(),
        "[sif_macro] A case has an insufficient amount of arguments ({} parameter(s) registered, but {} argument(s) were supplied)",
        params.len(),
        exprs.len()
    );

    let bindings = (0..params.len()).map(|i| create_binding(params[i], exprs[i]));

    quote! {
        #[test]
        #(#attributes)*
        #vis fn #ident() {
            #(#bindings)*
            #body
        }
    }
}

fn create_binding(param: (&Ident, &Type), expr: &Expr) -> proc_macro2::TokenStream {
    let (ident, typ) = param;

    quote! {
        let #ident: #typ = #expr;
    }
}
