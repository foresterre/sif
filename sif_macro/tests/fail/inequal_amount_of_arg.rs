use sif_macro::parameterized;

#[parameterized]
#[case("hello", 1)]
#[case("hello")]
pub(crate) fn my_test(v: &str, w: i32) {}

fn main() {}
