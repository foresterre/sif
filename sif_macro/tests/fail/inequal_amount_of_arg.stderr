error: custom attribute panicked
 --> $DIR/inequal_amount_of_arg.rs:3:1
  |
3 | #[parameterized]
  | ^^^^^^^^^^^^^^^^
  |
  = help: message: assertion failed: `(left == right)`
            left: `2`,
           right: `1`: [sif_macro] A case has an insufficient amount of arguments (2 parameter(s) registered, but 1 argument(s) were supplied)
